### DISCUSSION TITLE
re-use the temp text in the cache to speed up

### DISCUSSION THREAD

**Discussion by ragebear00** [dnGrep/dnGrep#1155] [Ideas]
Need to repeat searching a folder with 300K files. During the search, the target PDF/DOCX/EXCEL can be converted to text file in the temp folder.  

Can those temp text saved for future use?
-  if the PDF/DOCX/EXCEL temp text file is found and date of temp text file is later than the date of PDF/DOCX/EXCEL, OR 
- add MD5 at the file name. if the MD5 is the same
the  dnGREP will not convert again but re-use those temp text directly. 

The user can choose to keep temp text OR empty the temp text when exit dnGREP.  

It can increase speed dramatically.

**Comment by doug24** [comment]
Yes, I'm sure this would be beneficial when searching so many files (I never thought that someone would do that many).  I'll put this on the list of things to do, but not sure when I will get to it.  There needs to be a mapping between the original file and the temp text file, and currently that is stored in the search result.  This will now have to be stored somewhere else, and persisted to file if it the temp files are not deleted at the end of each session.

**Reply by ragebear00** [reply]
MD5 of the original file could be link. A new fresh launched research can create the map again between original file and temp text file again according to the MD5, since every file shall have a unique MD5. 

In extreme cases, the name of the temp text can be MD5 only and do not include string of the original filename. Once the MD5 of the original file created, dnGREP will check the existing MD5 in the temp folder at first. If it has been there, map the temp text to this original file, and skip convert. 

if the cached text can be stored in a ZIP/7z file, it can further save the space on the disk.

**Reply by doug24** [reply]
This is a good idea.  Using a file hash would mean there would not have to be any temporary or persistent mapping, or any need to check the file timestamp.  Take the hash of the original file and create the text file name using the original file name, underscore, hash, and .txt. extension. On search, take the hash of the original pdf/docx/xlsx file, construct the text file name and check if it already exists.  

`style-guide.pdf_19BE63329581CB3DF5258B496CE089685DD7848324DF7AE9D87E2C781FB1E5D9.txt`

I did a benchmark with a 9581 KB PDF file, and SHA256 is twice as fast as MD5.  It does add 64 characters to the file name (or 44 in Base64) instead of 32, but I don't think that matters as much as performance.  Not sure if there is any advantage to not including the original file name?  Having the file name makes it much easier to debug. The text files get squashed into a single directory for each file type, the temp files do not retain the original directory structure.

dnGrep does need the text file uncompressed for the preview pane, so cycling between zipped and unzipped files seems like a lot of overhead.  Maybe on application exit it could zip the entire directory of temp files and unzip it again on first use. 

I'm not going to have time to implement this for the July release (tomorrow) but can do it for August.

**Comment by ragebear00** [comment]
_Not sure if there is any advantage to not including the original file name?_ 

the advantage I believe, 
- the original filename can be a long string, and add the hash string in the name can hit the 255 limit in the MS system (the whole path + name shall not exceed 255 characters).  We may use part of the original filename (such as the beginning 10 characters) to limit the length of the new name
- some files are rename without any changes. If the hash value the same, we may refer to both of them. Of course, it may cause other issues. I don't know whether it is good or bad.

**Comment by ragebear00** [comment]
Agreed, ZIP/UNZIP is not useful then. If the disk is full, the unzip will make it messy.

If the cache folder can be a user selected location could be more practical solution.
